controllers/databasecluster_controller.go

```diff
@@ -1050,45 +1146,129 @@ func (r *DatabaseClusterReconciler) reconcilePXC(ctx context.Context, req ctrl.R
+			"s3.conf": pgbackrestS3Conf,
 		},
 	}
-	err = controllerutil.SetControllerReference(database, pgBackrestSecret, r.Scheme)
+	err := controllerutil.SetControllerReference(database, pgBackrestSecret, scheme)
 	if err != nil {
 		return nil, err
 	}
-	err = r.createOrUpdate(ctx, pgBackrestSecret, false)
+	err = CreateOrUpdate(ctx, client, pgBackrestSecret, false)
 	if err != nil {
 		return nil, err
 	}
 
 	return pgBackrestSecret, nil
 }
 
+// Returns the index of the backup storage in the pgbackrest repo list
+func GetBackupStorageIndexInPGBackrestRepo(
+	backupStorage *everestv1alpha1.BackupStorage,
+	repos []crunchyv1beta1.PGBackRestRepo,
+) int {
+	for idx, repo := range repos {
+		if repo.S3 == nil || repo.S3.Bucket != backupStorage.Spec.Bucket ||
+			repo.S3.Region != backupStorage.Spec.Region ||
+			repo.S3.Endpoint != backupStorage.Spec.EndpointURL {
+			continue
+		}
+		return idx
+	}
+
+	return -1
+}
+
+// Generates a PGBackrest repo for a given backup storage
+func GenPGBackrestRepo(
+	repoIdx int,
+	backupStorage *everestv1alpha1.BackupStorage,
+	backupSchedule *string,
+) (crunchyv1beta1.PGBackRestRepo, error) {
+	if repoIdx < 0 || repoIdx > 4 {
+		return crunchyv1beta1.PGBackRestRepo{}, errors.Errorf("repo index must be between 0 and 4")
+	}
+
+	pgRepo := crunchyv1beta1.PGBackRestRepo{
+		Name: fmt.Sprintf("repo%d", repoIdx),
+	}
+
+	if backupSchedule != nil {
+		pgRepo.BackupSchedules = &crunchyv1beta1.PGBackRestBackupSchedules{
+			Full: backupSchedule,
+		}
+	}
+
+	switch backupStorage.Spec.Type {
+	case everestv1alpha1.BackupStorageTypeS3:
+		pgRepo.S3 = &crunchyv1beta1.RepoS3{
+			Bucket:   backupStorage.Spec.Bucket,
+			Region:   backupStorage.Spec.Region,
+			Endpoint: backupStorage.Spec.EndpointURL,
+		}
+
+	default:
+		return crunchyv1beta1.PGBackRestRepo{}, errors.Errorf("unsupported backup storage type %s for %s", backupStorage.Spec.Type, backupStorage.Name)
+	}
+
+	return pgRepo, nil
+}
+
+// Adds the backup storage credentials to the PGBackrest secret data
+func AddBackupStorageCredentialsToPGBackrestSecret(
+	ctx context.Context,
+	client client.Client,
+	backupStorage *everestv1alpha1.BackupStorage,
+	repoName string,
+	data []byte,
+) ([]byte, error) {
+	cfg, err := ini.Load(data)
+	if err != nil {
+		return data, err
+	}
+
+	s3StorageSecret := &corev1.Secret{}
+	err = client.Get(ctx, types.NamespacedName{Name: backupStorage.Spec.CredentialsSecretName, Namespace: backupStorage.Namespace}, s3StorageSecret)
+	if err != nil {
+		return data, err
+	}
+
+	_, err = cfg.Section("global").NewKey(
+		fmt.Sprintf("%s-s3-key", repoName),
+		string(s3StorageSecret.Data["AWS_ACCESS_KEY_ID"]),
+	)
+	if err != nil {
+		return data, err
+	}
+
+	_, err = cfg.Section("global").NewKey(
+		fmt.Sprintf("%s-s3-key-secret", repoName),
+		string(s3StorageSecret.Data["AWS_SECRET_ACCESS_KEY"]),
+	)
+	if err != nil {
+		return data, err
+	}
+
+	buf := new(bytes.Buffer)
+	cfg.WriteTo(buf)
+	return buf.Bytes(), nil
```

Comment:
This seems to be an internal function.
Can we, instead of passing around `[]byte`, return `*ini.File` and marshal this to `[]byte`` only when needed in the spec?

This will save us from loading/writing bytes.